Added Dependency class with allow_failures (continued)

caffeinatedMike opened the pull request:
Applies logic that was started in #1611, fixed a few typos, and added additional logic to handle the new `Dependency` class.

Tests have been updated/added to test functionality ~~, but docs have not been updated~~.

**Edit:** Docs have now been updated.

Closes #1224 
Closes #1608 
Closes #1611

caffeinatedMike commented:
@selwin Everything seems to be in order according to the logic discussed in [your comment](https://github.com/rq/rq/issues/1608#issuecomment-1013542501) from #1608. However, `test_job_dependency` is failing for the newly-added cases that use `Dependency`. The dependent job seems to be stuck in the `DEFERRED` status.

Could you have a look?

selwin commented on rq/job.py:
```diff
@@ -988,6 +1016,7 @@ def dependencies_are_met(self, exclude_job_id=None, pipeline=None):
 
         return all(
             status.decode() == JobStatus.FINISHED
+            or (status.decode() in (JobStatus.FAILED, JobStatus.STOPPED) and self.allow_failure)
```
Can we make in `if` statement here? This is becoming hard to read.

selwin commented on rq/job.py:
```diff
@@ -42,6 +42,21 @@ class JobStatus(str, Enum):
+            raise ValueError("jobs: cannot be empty.")
+
+        self.dependencies = jobs
+        self.allow_failure = allow_failure
```
Can we make this into an enum to accommodate future variations? Some possible variations:
1. Execute immediately regardless of retries, or wait until retries are done
2. If a job has multiple dependencies, execute immediately after any one is successful3. 

Something like `Dependency(job, execute_when=EXECUTE_WHEN.SUCCESS)` or something like that (but with better naming possibly)?

selwin replied:
A few things that may come into consideration with regards to dependency logic:
1. If it has multiple dependencies
2. Retries

We don't need to address these in this PR, but preferably we design the dependency API so that it's more future proof.

caffeinatedMike commented:
@selwin can you please address my previous comment before I proceed to make these requested changes?

selwin commented:
@caffeinatedMike I just looked at the tests results. Many tests are failing so I think those will need to be fixed first. https://github.com/rq/rq/runs/5432544475?check_suite_focus=true

If you can make sure that existing test cases pass, I can then help you debug why the new code paths don't work. The comments I made in [this issue](https://github.com/rq/rq/issues/1608#issuecomment-1013542501) are just general guidelines, there could well be other implementation details that I missed.

caffeinatedMike commented:
@selwin I'm not sure where to start with resolving these failed tests because it seems many are unrelated to the code changes. From what I saw in other issues there are systemically-failing tests prior to my changes.

selwin commented:
Your [other pull request](https://github.com/rq/rq/pull/1629/files) seems to have passing tests. Could it be that you're basing your work off other branches that have test failures?

caffeinatedMike commented:
@selwin Actually it looks like the majority of tests didn't run because python 3.5 tests [failed](https://github.com/rq/rq/runs/5432541797?check_suite_focus=true) (due to use of f-string).

selwin commented:
Good catch, do you mind removing the usage of f string from your other PR?

Win~


On Mar 5, 2022 at 21:58:44, Michael Hill ***@***.***> wrote:

> @selwin <https://github.com/selwin> Actually it looks like the majority
> of tests didn't run because python 3.5 tests failed
> <https://github.com/rq/rq/runs/5432541797?check_suite_focus=true> (due to
> use of f-string).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rq/rq/pull/1630#issuecomment-1059778812>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABOB4XXMWJBKK4UJYQLFXTU6NZCJANCNFSM5PVWKICQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

caffeinatedMike commented:
@selwin I have the Dependency functionality working now with all-but-one test passing (`tests/test_job.py::TestJob::test_create_and_cancel_job_enqueue_dependents_in_registry`). Could you please run the workflows to verify this and assist in figuring out the final failure? Also, I'm not sure what additional changes you want (enum aside).

selwin commented:
@caffeinatedMike the tests are running. In the meantime, can we try to get https://github.com/rq/rq/pull/1629 merged first?

caffeinatedMike commented:
@selwin Shoot, I just pushed a commit that fixes the issue that causes that test to fail. As for #1629, unfortunately I'm running into some issues. Since `threading.Timer` is subclass of a `Thread` the exception on timeout occurs outside of the main thread, making it uncatchable (failure test fails even though the exception is thrown). I'm much rather focus on this major feature before circling back to that one.

caffeinatedMike commented:
@selwin my latest change [011c1f3](https://github.com/rq/rq/pull/1630/commits/011c1f3e1f0bf992566779b49101bec610fc1687) takes care of the last failing test. Please re-run the workflows. 

As for the enum, should this wait since we aren't sure what other variations might be needed? I'm hoping this feature can be merged asap.

caffeinatedMike commented:
@selwin my latest change [011c1f3](https://github.com/rq/rq/pull/1630/commits/011c1f3e1f0bf992566779b49101bec610fc1687) takes care of the last failing test. Please re-run the workflows. 

As for the enum, should this wait since we aren't sure what other variations might be needed? I'm hoping this feature can be merged asap.

selwin commented:
> As for the enum, should this wait since we aren't sure what other variations might be needed? I'm hoping this feature can be merged asap.

Ok, let's try to get this one in first.

selwin commented on docs/docs/index.md:
```diff
@@ -155,9 +155,33 @@ baz_job = queue.enqueue(baz, depends_on=[foo_job, bar_job])
+from rq import Queue
+
+queue = Queue(connection=Redis())
+job_1 = queue.enqueue(div_by_zero, on_success=report_success, on_failure=report_failure)
```
Can we remove the `on_success` and `on_failure` arguments here? They're not relevant in this context.

selwin commented on tests/test_worker.py:
```diff
@@ -700,21 +700,58 @@ def test_worker_sets_job_status(self):
+
+        # only enqueue dependent job when all dependencies have finished/failed
+        first_parent_job = q.enqueue(div_by_zero)
+        second_parent_job = q.enqueue(long_running_job, args=(5,))
```
Can we avoid using `long_running_job` for test cases where they're not needed? This will make tests run for longer.

caffeinatedMike replied:
What should this be replaced with here then?

selwin commented on rq/queue.py:
```diff
@@ -635,7 +638,9 @@ def enqueue_dependents(self, job, pipeline=None):
 
-                pipe.multi()
+                # only invoke .multi() if all dependent jobs' allow_failure are False-y
+                if not all([job.allow_failure for job in jobs_to_enqueue]):
+                    pipe.multi()
```
Sorry, I'm not sure I understand why this is needed.

caffeinatedMike replied:
Calling `pipe.multi()` here when any of the `jobs_to_enqueue` has `allow_failure=True` will cause `test_worker.py::TestWorker::test_job_dependency` to fail.

selwin replied:
I think this may be caused by the fact that there are [other commands](https://github.com/rq/rq/pull/1630/files#diff-50984f276e8a6fea5f4ba4938afb6f4fbc544fd3751fbd16027dc0a326d6405eR987) in the pipeline that are not yet executed.

I think we can simplify this by ensuring that `enqueue_dependents` are always called in a clean pipeline. Can you call `pipeline.execute()` prior to calling `enqueue_dependents`? This will make the logic easier to follow and troubleshoot.

caffeinatedMike replied:
Sure thing, will make this change and see how it does Monday.

caffeinatedMike replied:
@selwin I attempted this change (added a `pipeline.execute()` immediately before `queue.enqueue_dependents(job, pipeline=pipeline)` and removed the conditional statement), but it didn't not behave as expected (job still ends up as deferred). So far, the only solution was to conditionally execute the `pipe.multi()` command. Thoughts?

selwin replied:
Did you put `pipeline.execute()` before or after `pipeline.watch()`? I think it has to be called before `watch()`. https://github.com/redis/redis-py#pipelines

selwin replied:
The other alternative here is to call `enqueue_dependents` without `pipeline` so it can call `watch()` on its own (since there will always be two round trips to Redis anyway).

caffeinatedMike replied:
> Did you put `pipeline.execute()` before or after `pipeline.watch()`? I think it has to be called before `watch()`. https://github.com/redis/redis-py#pipelines
> The other alternative here is to call enqueue_dependents without pipeline so it can call watch() on its own (since there will always be two round trips to Redis anyway).

I've tried both ways (with removing the conditional statement around the `pipe.multi()` call in `enqueue_dependents`):
```python3
with suppress(redis.exceptions.ConnectionError):
    pipeline.execute()
# if dependencies are inserted after enqueue_dependents
# a WatchError is thrown by execute()
pipeline.watch(job.dependents_key)
# enqueue_dependents calls multi() on the pipeline!
queue.enqueue_dependents(job, pipeline=pipeline)
```
And
```python3
with suppress(redis.exceptions.ConnectionError):
    pipeline.execute()
# if dependencies are inserted after enqueue_dependents
# a WatchError is thrown by execute()
pipeline.watch(job.dependents_key)
# enqueue_dependents calls multi() on the pipeline!
queue.enqueue_dependents(job)
```

I think I'm going to have to call it quits on my efforts soon because I've already spent too much time trying to fix the dependency issues with this package. I'm sadly probably going to have to move on to another package.

selwin replied:
@caffeinatedMike no worries, thanks for your efforts!

selwin replied:
I'll try to look into this when I can, am super busy now.

selwin commented on rq/worker.py:
```diff
@@ -974,6 +977,12 @@ def handle_job_failure(self, job, queue, started_job_registry=None,
                                                         job_class=self.job_class, serializer=job.serializer)
                 failed_job_registry.add(job, ttl=job.failure_ttl,
                                         exc_string=exc_string, pipeline=pipeline)
+                try:
+                    pipeline.execute()
+                except Exception:
+                    # Ensure that custom exception handlers are called
+                    # even if Redis is down
+                    pass
```
Can we remove this code block? This is not relevant for this PR and I'm not sure calling `pipeline.execute()` here is the right thing to do.

caffeinatedMike replied:
This is required to make `test_job.py::TestJob::test_create_and_cancel_job_enqueue_dependents_in_registry` pass, because originally the job status would be stuck in the `started` state (maybe due to the addition of unconditionally enqueueing dependents below that code block in the same function)

caffeinatedMike commented:
@selwin 

- Addressed (made changes regarding) 2/4 of the latest round of review comments.
- For the other 2, I've responded with clarification on the reasons for the additional logic. Let me know if you need any further clarification or would like to offer an alternative approach.
- Added an additional test to improve coverage.

All checks are now passing 🎉

selwin commented on rq/worker.py:
```diff
@@ -974,6 +977,12 @@ def handle_job_failure(self, job, queue, started_job_registry=None,
                 failed_job_registry.add(job, ttl=job.failure_ttl,
                                         exc_string=exc_string, pipeline=pipeline)
+                try:
+                    pipeline.execute()
```
Can we change this to use `with suppress(Exception)`? It will give us better test coverage since the `pass` line will not be tested.

Also, is there a specific exception that covers Redis connection errors? Just so that we don't accidentally swallow other classes of exceptions.

caffeinatedMike replied:
Change completed. Logic now suppresses on redis connection errors with `with suppress(redis.exceptions.ConnectionError):`

selwin commented:
Superseded by https://github.com/rq/rq/pull/1681
